rust/src/accumulator.rs

```diff
@@ -0,0 +1,224 @@
+//! This module contains the structure that accumulates usage data
+//! locally before flushing to Kafka.
+//!
+//! The accumulator pre-aggregates usage per timestamp based on
+//! the granularity provided at instantiation.
+//!
+
+use chrono::{DateTime, Duration, DurationRound, Local};
+use serde::Serialize;
+use std::collections::HashMap;
+use std::fmt;
+use std::mem;
+
+/// The unit of measures we support when recording usage.
+/// more can be added.
+#[derive(Clone, Debug, Eq, Hash, PartialEq, Serialize)]
+#[serde(rename_all = "snake_case")]
+pub enum UsageUnit {
+    Milliseconds,
+    Bytes,
+    BytesSec,
+}
+
+impl fmt::Display for UsageUnit {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        match self {
+            UsageUnit::Milliseconds => write!(f, "milliseconds"),
+            UsageUnit::Bytes => write!(f, "bytes"),
+            UsageUnit::BytesSec => write!(f, "bytes_sec"),
+        }
+    }
+}
+
+#[derive(Clone, Debug, Eq, PartialEq, Hash)]
+pub struct UsageKey {
+    pub quantized_timestamp: DateTime<Local>,
+    pub resource_id: String,
+    pub app_feature: String,
+    pub unit: UsageUnit,
+}
+
+pub struct UsageAccumulator {
```

Comment:
remind me - is this supposed to be public again?

Reply:
Accessible within the Crate. Not from outside